### PR TITLE
Indra db

### PR DESCRIPTION
**Pull request by pagreene**
Merge indra db with db access and initial upload of pubmed, and most of pmc oa.

### CODE REVIEW

File: indra/db/__init__.py [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+@compiles(DropTable, "postgresql")
+def _compile_drop_table(element, compiler, **kwargs):
+    return compiler.visit_drop_table(element) + " CASCADE"
+
+try:
+    from pgcopy import CopyManager
+    CAN_COPY = True
+except:
+    print("WARNING: pgcopy unavailable. Bulk copies will be slow.")
+    CopyManager = None
```
Comment (johnbachman) [inline indra/db/__init__.py]: Use logger instead of print.

File: indra/db/__init__.py [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+                query_args = [self.tables[tbls]]
+            else:
+                raise InputError('Unrecognized table specification type: %s.' %
+                                 type(tbls))
+                
+        return self.session.query(*query_args).filter(*args)
+
+
+    def select_one(self, tbls, *args):
+        """Select the first value that matches requirements given in kwargs
```
Comment (johnbachman) [inline indra/db/__init__.py]: Single line top-level docstring.

File: indra/db/__init__.py [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+        """Select the first value that matches requirements given in kwargs 
+        from table indicated by tbl_name.
+        
+        Note that if your specification yields multiple results, this method 
+        will just return the first result without exception.
+        """
+        return self.filter_query(tbls, *args).first()
+
+
+    def select(self, tbls, *args):
```
Comment (johnbachman) [inline indra/db/__init__.py]: Change to select_all

File: indra/db/__init__.py [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+                            db_name=db_name, 
+                            db_id=db_id
+                            )
+                        )
+                self.insert_many('agents', input_list)
+        return
+
+
+    def get_abstracts_by_pmids(self, pmid_list, unzip=True):
+        "Get abstracts using teh pmids in pmid_list."
```
Comment (johnbachman) [inline indra/db/__init__.py]: teh lolz

File: indra/db/build_sample_set.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
+from indra.literature import pubmed_client as pub
```
Comment (johnbachman) [inline indra/db/build_sample_set.py]: Add from future import unicode_literals, etc.

File: indra/db/build_sample_set.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+def _get_example(case, med_pmid_list, pmc_dicts, man_dicts):
+    if case == (1, 0, 0):
+        pmid = random.choice(med_pmid_list)
+        pmc_pmid_list = [d['PMID'] for d in pmc_dicts if d['PMID'] != '']
+        # pmc has only a tiny fraction of all pmid's, so this will be very fast.
+        while pmid in pmc_pmid_list:
+            pmid = random.choice(med_pmid_list)
+        ret = (pmid, '', '')
+    elif case == (1, 1, 0):
+        pmc_pmid_dict_list = [d for d in pmc_dicts if d['PMID'] != '']
```
Comment (johnbachman) [inline indra/db/build_sample_set.py]: If you feel like it, refactor to use set intersections

File: indra/db/build_sample_set.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    '''
+    Create the nastiest set of content we're willing/able to handle.
+    
+    We create a small local representation of the entirety of the nih 
+    repositories we use, including all the nasty corner cases we can manage.
+    This allows for rapid development and testing.
+    
+    Inputs
+    ------
+    n: int
```
Comment (johnbachman) [inline indra/db/build_sample_set.py]: Space before colon (check readthedocs after merging)

File: indra/db/build_sample_set.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        with tarfile.open(dirname + '.tar.gz', 'w:gz') as tar:
+            tar.add(dirname)
+        shutil.rmtree(dirname)
+    
+    return examples
+
+if __name__ == '__main__':
+    from sys import argv
+    assert len(argv) == 3, "Error with args: %s." % argv
+    n = int(argv[1])
```
Comment (johnbachman) [inline indra/db/build_sample_set.py]: Add a few comments about what the command-line args are.

File: indra/db/populate_content.py [hunk 9]
```diff
@@ -0,0 +1,705 @@
...
+
+ftp_blocksize = 33554432 # Chunk size recommended by NCBI
+BATCH_SIZE = 10000
+
+    
+
+
+class NihFtpClient(object):
+    '''
+    High level access to the nih ftp repositories.
```
Comment (johnbachman) [inline indra/db/populate_content.py]: Capitalize acronyms

File: indra/db/populate_content.py [hunk 10]
```diff
@@ -0,0 +1,705 @@
...
+    
+    Parameters
+    ----------
+    my_path: str 
+        The path to the subdirectory around which this client operates.
+    ftp_url: str: (default: 'ftp.ncbi.nlm.nih.gov')
```
Comment (johnbachman) [inline indra/db/populate_content.py]: For optional arguments, `Optional[str]`. Include default values at the end of the arg description.

File: indra/db/populate_content.py [hunk 11]
```diff
@@ -0,0 +1,705 @@
...
+                buf.flush()
+        else:
+            with open(self.path_join(self.ftp_url, full_path), 'rb') as f:
+                buf.write(f.read())
+                buf.flush()
+        return
+
+
+    def download_file(self, f_path):
+        name = path.basename(f_path)
```
Comment (johnbachman) [inline indra/db/populate_content.py]: Document differences between download_file, get_file, and ret_file.

File: indra/db/populate_content.py [hunk 12]
```diff
@@ -0,0 +1,705 @@
...
+            with self.get_ftp_connection() as ftp:
+                contents = ftp.nlst()
+        else:
+            contents = os.listdir(self.path_join(self.ftp_url, ftp_path))
+        return contents
+
+
+class Uploader(object):
+    '''
+    This abstract class provides the api required for any object that is used
```
Comment (johnbachman) [inline indra/db/populate_content.py]: Top level 80 char docstring.

File: indra/db/populate_content.py [hunk 13]
```diff
@@ -0,0 +1,705 @@
...
+            )
+    
+    def update(self, db):
+        "A stub for the method used to update the content on the database."
+        raise NotImplementedError(
+            "`Update` not implemented for `%s`." % self.__class__.__name__
+            )
+
+class NihUploader(Uploader):
+    '''
```
Comment (johnbachman) [inline indra/db/populate_content.py]: One-liner docstring.

File: indra/db/populate_content.py [hunk 14]
```diff
@@ -0,0 +1,705 @@
...
+    def fix_doi(self, doi):
+        "Sometimes the doi is doubled (no idea why). Fix it."
+        if doi is None:
+            return doi
+        L = len(doi)
+        if L%2 is not 0:
+            return
+        if doi[:L] != doi[L:]:
+            return
+        print("\tFixing doubled doi: %s" % doi)
```
Comment (johnbachman) [inline indra/db/populate_content.py]: Change prints to logging

File: indra/db/populate_content.py [hunk 15]
```diff
@@ -0,0 +1,705 @@
...
+        n_procs: int
+            The number of processes to use when parsing xmls.
+        continuing: bool
+            If true, assume that we are picking up after an error, or otherwise
+            continuing from an earlier process. This means we will skip over
+            source files contained in the database. If false, all files will be
+            read and parsed.
+        """
+        xml_files = self.get_file_list()
+        tr_list = db.select(
```
Comment (johnbachman) [inline indra/db/populate_content.py]: sf_list?

File: indra/db/populate_content.py [hunk 16]
```diff
@@ -0,0 +1,705 @@
...
+            )
+        existing_files = [tr.name for tr in tr_list]
+        
+        q = mp.Queue()
+        proc_list = []
+        for xml_file in xml_files:
+            if continuing and xml_file in existing_files:
+                print("Skipping %s. Already uploaded." % xml_file)
+                continue
+            if xml_file not in existing_files:
```
Comment (johnbachman) [inline indra/db/populate_content.py]: Refactor to handle issue of incomplete upload due to error in file already marked as uploaded.

File: indra/db/populate_content.py [hunk 17]
```diff
@@ -0,0 +1,705 @@
...
+                    else:
+                        tr_entry['pmid'] = pmid
+                        add_record(tr_entry)
+                elif tr_entry['pmid'] in db_conts['pmid']:
+                    if get_tr_from_pmid(tr_entry['pmid']).pmcid is None:
+                        update_record_with_pmcid(tr_entry)
+                    else:
+                        with open('review.txt', 'a') as f:
+                            f.write(
+                                'Skipped record from %s with pmcid %s due to conflicting pmids. Please review.\n'
```
Comment (johnbachman) [inline indra/db/populate_content.py]: Check line length

File: indra/db/populate_content.py [hunk 18]
```diff
@@ -0,0 +1,705 @@
...
+                                )
+                        pmcids_to_skip.append(tr_entry['pmcid'])
+                else:
+                    add_record(tr_entry)
+        print("\nDone filtering text refs...")
+        return filtered_tr_records, pmcids_to_skip
+
+
+    def filter_text_content(self, db, tc_data):
+        'Filter the text content'
```
Comment (johnbachman) [inline indra/db/populate_content.py]: Doc: filter text content to identify pre-existing records, conflicts

File: indra/tests/test_springer.py [hunk 19]
```diff
@@ -0,0 +1,167 @@
+from __future__ import absolute_import, print_function, unicode_literals
```
Comment (johnbachman) [inline indra/tests/test_springer.py]: Move springer-related files/content to another branch for separate PR.

**Comment by bgyori** [comment]
Just noticed a problem, namely, since `db` is inside `indra`, `nosetests --with-doctest` tries to load all of its files on Travis. But it errors in the Python 2 build, so we need to find a way to work around this.

**Comment by pagreene** [comment]
I feel like raise an import error is the appropriate response...I suppose I
could just log a warning saying this is not supported for python 2, but
then I think a bunch of the imports will fail because of naming
differences. It would be another can of worms to make it python2 compatible.

It looks like there is a '-I' (dash - eye) option for nostests to ignore
certain files.

On Wed, Sep 27, 2017 at 5:22 PM, Benjamin M. Gyori <notifications@github.com

**Comment by bgyori** [comment]
Okay, build is passing for me now!

**Comment by johnbachman** [comment]
Nice!

**Pull request merged**